src/gleam/time/timestamp.gleam

```diff
@@ -227,12 +265,388 @@ fn to_civil(minutes: Int) -> #(Int, Int, Int) {
+            count:,
+            acc: acc * 10 + { byte - 0x30 },
+            k: k + 1,
+          )
+        _ -> Error(Nil)
+      }
+    }
+    order.Gt | order.Eq -> Ok(#(acc, bytes))
+  }
+}
+
+/// Accept the given value from `bytes` and move past it if found.
+/// 
+fn accept_byte(from bytes: BitArray, value value: Int) -> Result(BitArray, Nil) {
+  case bytes {
+    <<byte, remaining_bytes:bytes>> if byte == value -> Ok(remaining_bytes)
+    _ -> Error(Nil)
+  }
+}
+
+fn accept_date_time_separator(from bytes: BitArray) -> Result(BitArray, Nil) {
+  case bytes {
+    <<byte, remaining_bytes:bytes>>
+      if byte == byte_t_uppercase || byte == byte_t_lowercase
+    -> Ok(remaining_bytes)
+    _ -> Error(Nil)
+  }
+}
+
+fn accept_empty(from bytes: BitArray) -> Result(Nil, Nil) {
+  case bytes {
+    <<>> -> Ok(Nil)
+    _ -> Error(Nil)
+  }
+}
+
+/// Note: The caller of this function must ensure that all inputs are valid.
+/// 
+fn from_date_time(
+  year year: Int,
+  month month: Int,
+  day day: Int,
+  hours hours: Int,
+  minutes minutes: Int,
+  seconds seconds: Int,
+  second_fraction_as_nanoseconds second_fraction_as_nanoseconds: Int,
+  offset_seconds offset_seconds: Int,
+) -> Timestamp {
+  let julian_seconds =
+    julian_seconds_from_parts(year:, month:, day:, hours:, minutes:, seconds:)
+
+  let julian_seconds_since_epoch = julian_seconds - julian_seconds_unix_epoch
+
+  let timestamp =
+    Timestamp(
+      seconds: julian_seconds_since_epoch,
+      nanoseconds: second_fraction_as_nanoseconds,
+    )
+    |> normalise
+
+  let offset = Timestamp(seconds: offset_seconds, nanoseconds: 0)
+  subtract(timestamp, offset)
+}
+
+/// `julian_seconds_from_parts(year, month, day, hours, minutes, seconds)` 
+/// returns the number of Julian 
+/// seconds represented by the given arguments.
+/// 
+/// Note: It is the callers responsibility to ensure the inputs are valid.
+/// 
+/// See https://www.tondering.dk/claus/cal/julperiod.php#formula
+/// 
+fn julian_seconds_from_parts(
+  year year: Int,
+  month month: Int,
+  day day: Int,
+  hours hours: Int,
+  minutes minutes: Int,
+  seconds seconds: Int,
+) {
+  let julian_day_seconds =
+    julian_day_from_ymd(year:, month:, day:) * seconds_per_day
+
+  julian_day_seconds
+  + { hours * seconds_per_hour }
+  + { minutes * seconds_per_minute }
+  + seconds
+}
+
+/// Note: It is the callers responsibility to ensure the inputs are valid.
+/// 
+/// See https://www.tondering.dk/claus/cal/julperiod.php#formula
+/// 
+fn julian_day_from_ymd(year year: Int, month month: Int, day day: Int) -> Int {
+  let adjustment = { 14 - month } / 12
+  let adjusted_year = year + 4800 - adjustment
+  let adjusted_month = month + 12 * adjustment - 3
+
+  day
+  + { { 153 * adjusted_month } + 2 }
+  / 5
+  + 365
+  * adjusted_year
+  + { adjusted_year / 4 }
+  - { adjusted_year / 100 }
+  + { adjusted_year / 400 }
+  - 32_045
+}
+
+fn subtract(left: Timestamp, right: Timestamp) -> Timestamp {
```

Comment:
This function doesn't make sense. You cannot subtract a timestamp from a timestamp.